When using (*JsonParseNode).GetXValue methods it panics fix

michaeldcanady opened the pull request:
Added utils for converting from like types to others:
ex:
- int8 -> float64
- float32 -> float64
- float64 -> int8 (as long as it doesn't exceed int8 max and/or have a decimal place)
doesn't convert between non-like types:
bool(true) -/-> "true"
int8(1) -/-> bool(true)

Fixes #142

baywet commented on util.go:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	// Make sure nothing is a pointer
+	valValue := reflect.ValueOf(in)
```
while I understand the benefits of the approach for numeric types and make sure a conversion is possible/desirable, I'm a bit worried about the potential performance impact of using reflection for the other types.
Have we considered a hybrid approach here? where string/bool/etc... would rely on type assertions and numeric types would rely on this util method?

michaeldcanady replied:
just to make sure I am understanding, your recommendation would be to have only numeric types implement the as function and the "other" types only use type assertions?
ex:
numeric
```golang
func (n *JsonParseNode) GetInt8Value() (*int8, error) {
	if isNil(n) || isNil(n.value) {
		return nil, nil
	}
	var val int8

	if err := as(n.value, &val); err != nil {
		return nil, err
	}

	return &val, nil
}
```
string
```golang
func (n *JsonParseNode) GetStringValue() (*string, error) {
	if isNil(n) || isNil(n.value) {
		return nil, nil
	}
	
       val, ok := n.value.(string)
       if !ok {
          return nil, fmt.Errorf("value '%v' is not compatible with type string", n.value)
       }

	return &val, nil
}
```

michaeldcanady replied:
The error would need to be tweaked since it's a memory address but not really relevant.

baywet replied:
correct, that's the question I was asking.

michaeldcanady replied:
I don't see any reason not to except for a weird edge case, which shouldn't happen without user error, where the value is set to a double pointer or a string nested in an interface...?

michaeldcanady replied:
for example this would error:
```golang
func main() {

	str := "str"
	strPtr := &str

	node, _ := jsonserialization.NewJsonParseNode([]byte(`"dfsdfds"`))
	node.SetValue(&strPtr)

	val, err := node.GetStringValue()
	if err != nil {
		log.Fatal(err)
	}
	log.Println(val)
}
```
It's just a matter of how much you want to "protect the user" vs what is the performance impact. It might be worth testing it in a dev build, have an intensive object, and compare the two.

baywet replied:
Thanks for calling this out, this was an oversight on our end, SetValue should never have been public to begin with. I put together a PR to address this. https://github.com/microsoft/kiota-serialization-json-go/pull/145
Do you have other scenarios where you think the change I'm recommending could be problematic?

michaeldcanady replied:
Since `SetValue` will be effectively removed. I can't see any other concerns since only valid JSON will be imported. I can make the change for non-number types.

baywet reviewed (commented):
Thanks for the contribution! I did leave one question about the impact here.

michaeldcanady commented:
reopening, it closed, by mistake, when I went to resolve the conflicts

baywet reviewed (dismissed):
Thank you for making the changes!
Can you also add an entry to the changelog please? (patch bump, changed, today's date)
@andrueastman for final review and merge/publish

michaeldcanady commented:
> Thank you for making the changes! Can you also add an entry to the changelog please? (patch bump, changed, today's date) @andrueastman for final review and merge/publish

Just added, let me know if any adjustments should be made to the verbiage!

andrueastman reviewed (approved):
Thanks for the contribution here @michaeldcanady
